[Lollo] Step 4 - 포커 딜러와 게임 테스트

eeeesong opened the pull request:
Step 4를 완료하여 PR 보냅니다. 너무 어렵네요..🤯

### 주요 학습 내용

- **지난 스텝 리팩토링**
  - `Suit`과 `Rank` enum에도 `CustomStringConvertible`을 적용하고, `Suit`의 rawValue 표기와 `Rank`의 case 명칭을 변경하였습니다.
  - 메소드 순서, 줄바꿈 등을 수정하였습니다.
- **객체지향적으로 구현하기**
  - 처음엔 `PokerGame` class의 자식으로 `Dealer`가 있고, `Dealer`가 `Player` 타입 및 players 배열을 가지고 있는 것으로 구현하였습니다.
  - 그러나 `Dealer`에게 `Player`들이 속해있는 것이 어색해보여서 `Player`를 `PokerGame` class의 중첩 타입으로 이동시키고, 전반적인 수정을 거쳤습니다.
  - 또한, enum으로 선언한 `Rule`과 `Seat`의 **rawValue**를 통해 메소드 내부 연산을 하는 것으로 작성했다가 구체적인 의미를 나타낼 수 있도록 수정해보았습니다. 각각 cardCount(), count() 메소드를 따로 만들어 필요한 개수를 리턴하도록 했습니다.
  - 현재 `ViewController`에서 **dealer** 오브젝트를 생성하고, 그 오브젝트를 통해 게임을 시작하는 것으로 작성이 되어 있는데, **pokerGame** 오브젝트가 아닌 dealer를 통하더라도 괜찮은 것인지 의문이 있습니다.

eeeesong committed:
README: Step 3 구현 내용 정리

eeeesong committed:
Step 3 내용 리팩토링

- Card: Suit, Rank enum에 CustomStringConvertible 적용 / Rank case 명 변경
- CardDeck: 메소드 위치, 줄바꿈 수정

eeeesong committed:
PokerGame class 생성 및 기능 구현

eeeesong committed:
리팩토링

eeeesong committed:
기능 테스트에 맞추어 메소드 업데이트

- PrintFactory: endMessage() 추가
- ViewController: 테스트 코드 추가

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+
+class PokerGame: CustomStringConvertible {
+
+    var rule: Rule
```
항상 속성은 private var를 우선적으로 고려하세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+        case one = 1, two, three, four
+        
+        func count() -> Int {
+            return self.rawValue + 1 //딜러 몫 추가
```
enum 타입에 메소드를 추가한 시도는 매우 긍정적입니다. 다만 값을 그대로 전달해주는 게 아니라 
Rule.cardCount() 나 Seat.count() 값을 가져가서 어떻게 사용하는지 살펴보고 그 동작을 제공하려고 해보세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+    private func addPlayers() {
+        for i in 1..<seat.count() {
+            players.append(Player(name: "참가자 #\(i)"))
+        }
```
예를 들어 seat.count()를 가져와서 for 구문을 만들어서 반복하는 거라면 
`players.append(Player(name: "참가자 #\(i)"))`를 seat에 전달하고 반복을 seat 내부에서 처리하는 방법도 가능합니다.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+                status += "\(player.name): \(cards)\n"
+            }
+        }
+        return status
```
description을 제공하려는 시도는 좋습니다만,
description으로 모든 정보를 출력하려고 또는 문자열로 다 만들어서 넘겨주려는 것은 좋지 않습니다.
앱의 뷰에 문자열만 출력하는 게 아니니까요. 항상 데이터 자체를 전달할 수 있는지 고려해보세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+class Dealer: PokerGame {
```
Dealer 타입이 PokerGame을 상속하는 것은 많이 어색하네요.
가상의 포커 게임을 생각해봅시다. 딜러는 게임 그 자체인걸까요? 
CS10 OOP 수업에서 클래스를 어떤 기준으로 나눠야 하는지 언급했던 것을 찾아보세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    func start() -> String {
+        let cardNeeded = seat.count() * rule.cardCount()
+        guard cardDeck.count() >= cardNeeded else { return P.endMessage() }
```
이 부분도 cardDeck의 값을 가져와서 비교하기 보다는 `cardDeck.isAvailable(seat, rule)` 형태가 되면 어떨까요?
중요한 포인트는 호출하는 상위 모듈이 호출당하는 하위 모듈보다 간결해야 한다는 점입니다. 
객체나 특정 타입의 값을 가져와서 처리하는 코드는 항상 그 객체나 타입 내에서 처리하는 방식으로 개선할 수 있습니다.

godrm commented:
PR 열어놓을테니 피드백 내용을 확인해서 수정사항을 다시 PUSH해보세요.
모두 수정하고나면 댓글을 남겨주세요.

eeeesong commented:
넵! 피드백 확인했습니다
클래스 구조를 고심한 후 수정해보겠습니다 :)
감이 잘 잡히지 않아서 시간이 좀 걸릴 것 같습니다!!

eeeesong committed:
class 구조 변경 - Participant class 추가

- PokerGame에 속해 있던 Dealer와 Player 분리
- Participant를 상위 클래스로 두고 하위 클래스로 Dealer, Player 각각 정의

eeeesong committed:
class 구조 변화에 따른 사소한 수정

- ViewController: 테스트 코드 변경
- CardDeck: shuffle() 반환 값 삭제

eeeesong committed:
프린트 방식 개선을 위한 수정

- InfoBoard class를 따로 생성하여 현 상황을 업데이트 받도록 함
- String 외의 출력 방식도 가능하도록 설계

eeeesong committed:
PrintFactory 삭제

eeeesong committed:
리팩토링

eeeesong commented:
JK 안녕하세요! 좋은 아침입니다 :)
어제 말씀해주신 피드백 반영하여 수정 진행하였습니다.
수정 방향은 크게 두가지입니다.

1. **class 구조 변경**
  - `PokerGame`에 속해 있던 `Dealer`와 `Player`를 빼냈습니다.
  - 그리고 두 class를 `Participant`라는 class의 자식 class로 설정하여, 공통된 사항은 부모 class를 통할 수 있도록 작성하였습니다.
2. **게임 상황 저장 방식 개선**
  - 게임의 진행 상황을 일괄적으로 String으로 출력하던 방식을 개선했습니다.
  - `InfoBoard` class를 따로 만들고, 필요한 정보들을 각 타입에 맞춰 따로 저장하도록 했습니다.
  - 추후 View 통해 출력하기 위해 개별 정보가 필요할 때 쉽게 접근이 가능할 것으로 생각했습니다.
    - 다만, 현재 `PokerGame`과 `Dealer` 둘 다 해당 class 접근을 통해 정보를 업데이트하는데, 이 정보만 업데이트하는 객체를 따로 두는 것이 나은 지 고민이 됩니다.
  - 현재는 콘솔 창에 일괄 출력하는 것만 필요하므로, `CustomStringConvertible`을 통해 필요한 메시지를 반환할 수 있도록 했습니다.

godrm commented on PokerGameApp/PokerGameApp/Participant.swift:
```diff
@@ -0,0 +1,126 @@
+            case .player:
+                return "플레이어"
+            }
+        }
```
이렇게 Role 타입에 따라서 분기해서 처리하는 로직보다는 타입을 분리하는 것을 시도해보세요. 
아래 나눠져 있군요. 그러면 여기 처리도 각각 아래 타입에서 처리하면 됩니다.

godrm commented:
다음 단계를 진행하면서 수업 내용을 고려해서 개선해보세요.
